parachains/pallets/bridge-transfer/src/lib.rs

```diff
@@ -0,0 +1,1596 @@
+				AllowedReserveLocations::<T>::try_mutate(|locations| -> Result<bool, Error<T>> {
+					let mut removed = false;
+					for ltr in locations_to_remove {
+						let ltr: MultiLocation =
+							ltr.try_into().map_err(|_| Error::<T>::UnsupportedDestination)?;
+						removed |= locations.remove(&ltr);
+					}
+					Ok(removed)
+				})?;
+			if removed {
+				Self::deposit_event(Event::ReserveLocationRemoved);
+			}
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Validates destination and check if we support bridging to this remote global consensus
+		///
+		/// Returns: correct remote location, where we should be able to bridge
+		pub(crate) fn ensure_remote_destination(
+			remote_destination: VersionedMultiLocation,
+		) -> Result<(NetworkId, BridgeConfig, MultiLocation), Error<T>> {
+			match remote_destination {
+				VersionedMultiLocation::V3(remote_location) => {
+					ensure!(
+						remote_location.parent_count() == 2,
+						Error::<T>::UnsupportedDestination
+					);
+					let local_network = T::UniversalLocation::get()
+						.global_consensus()
+						.map_err(|_| Error::<T>::InvalidConfiguration)?;
+					let remote_network = remote_location
+						.interior()
+						.global_consensus()
+						.map_err(|_| Error::<T>::UnsupportedDestination)?;
+					ensure!(local_network != remote_network, Error::<T>::UnsupportedDestination);
+					match AllowedExporters::<T>::get(remote_network) {
+						Some(bridge_config) => {
+							ensure!(
+								// TODO:check-parameter - verify and prepare test for ETH scenario - https://github.com/paritytech/cumulus/pull/2013#discussion_r1094909290
+								remote_location.starts_with(&bridge_config.allowed_target_location),
+								Error::<T>::UnsupportedDestination
+							);
+							Ok((remote_network, bridge_config, remote_location))
+						},
+						None => return Err(Error::<T>::UnsupportedDestination),
+					}
+				},
+				_ => Err(Error::<T>::UnsupportedDestination),
+			}
+		}
+
+		#[transactional]
+		fn do_reserve_and_send_in_transaction(
+			origin_location: MultiLocation,
+			remote_destination: MultiLocation,
+			assets: MultiAssets,
+			bridge_config: BridgeConfig,
+		) -> Result<(), DispatchError> {
+			// Resolve reserve account as sovereign account of bridge
+			let reserve_account = bridge_config.bridge_location;
+
+			let allowed_target_location = bridge_config.allowed_target_location;
+
+			// lets try to do a reserve for all assets
+			let mut reserved_assets = xcm_executor::Assets::new();
+			for asset in assets.into_inner() {
+				// TODO:check-parameter - verify this Joe's text
+				// Deposit assets into `AccountId` that corresponds to the bridge
+				// hub. In this way, Statemine acts as a reserve location to the
+				// bridge, such that it need not trust any consensus system from
+				// `./Parent/Parent/...`. (It may trust Polkadot, but would
+				// Polkadot trust Kusama with its DOT?)
+
+				// Move asset to reserve account
+				T::AssetTransactor::transfer_asset(
```

Comment:
Reserve accounts are for assets that are native to the source chain. For an asset that is native to the target chain the reserve account will live on the target chain. If the asset is native to the target chain we should burn the asset instead of transferring it to a reserve account.

We will need to check if the asset is native here by looking at its MultiLocation and if it belongs to the same consensus system.

psuedo code:
```rust
if asset.is_native() {
    transfer_asset(asset, ...);
}
else {
    burn_asset(asset, ...);
}
reserved_assets.subsume_assets(asset);
```

Reply:
@alistair-singh 
thank you, yes, agree, actually this one of internal open points/todos, I have similar `IF` also in my mind, but giving it some time

maybe the `native` word is not exact, maybe `from the same global consensus`,
for example:
KSM is native for Kusama relay chain and also for Kusama/Statemine system parachain, statemine see KSM as `MultiLoation(parents: 1, Here)` but Statemint will see KSM as `MultiLoation(parents: 2, X1(GlobalConsensus(Kusama)))`,
or e.g. other asset on Statemine from TrustBackedAssets pallet instance, Statemint will see as `MultiLoation(parents: 2, X3(GlobalConsensus(Kusama), Parachain(1000),PalletInstance(50),GeneralIndex(12356)))`


also interesting it will be for scenario like ETH:

1.  ETH -> BridgeHubPolkadot -> Statemint (ends in ForeignAssets pallet_assets instance with id like `MultiLocation(parents: 2, GlobalConsensus(Ethereum(eth_chain_Id))) + maybe some smartcontract accountid?`
2.  ForeignAsset ETH bridge transfer from Polkadot/Statemint to Kusama/Statemine? allow this kind of transfer?

ETH here is hard to say, if it is native to Statmine or Statemint

Reply:
> maybe the `native` word is not exact, maybe `from the same global consensus`

Agreed. `ensure_is_remote` naming is used for this in [universal exports](https://github.com/paritytech/polkadot/blob/6dc9e84dde2f28215ab4e994cc021676d942eb12/xcm/xcm-builder/src/universal_exports.rs#L26). 

> 2\. ForeignAsset ETH bridge transfer from Polkadot/Statemint to Kusama/Statemine? allow this kind of transfer?

Yes, I see, this is complicated. We should not allow this type of round trip transfer. 

> ETH here is hard to say, if it is native to Statmine or Statemint

If we use the `ensure_is_remote` logic it would be remote to both Statemine and Statemint, and on both parachains we should use the burn logic. This is the right thing to do in both these cases.

Reply:
I used `ensure_is_remote` here https://github.com/paritytech/polkadot/pull/7016 and if I have approve there, I will use it here also no problem, but lets wait for that converter first

Reply:
Just circling back to this, we do require this change for our bridge: burning the asset, and sending a message like the following to Ethereum:
```
UniversalOrigin(/GlobalConsensus(Polkadot))
DescendOrigin(/Para(Statemint))
WithdrawAsset(Asset)
ClearOrigin
DepositAsset(Asset, Destination)
```

Reply:
> Just circling back to this, we do require this change for our bridge: burning the asset, and sending a message like the following back to Ethereum:
> 
> ```
> UniversalOrigin(/GlobalConsensus(Polkadot))
> DescendOrigin(/Para(Statemint))
> WithdrawAsset(Asset)
> ClearOrigin
> DepositAsset(Asset, Destination)
> ```

@vgeddes @alistair-singh 
yes, exactly, it is almost done, at least tests are working, but need to finish big configuration refactor, I am ooo and back on Monday, so then I will push everything